### PR TITLE
Define native Haskell stack-graph-path structure.

### PR DESCRIPTION
**Pull request by patrickt**
This defines the “native” Haskell data type that describes partial paths within a stack graph. Right now it only implements the type and functions to check its invariants. Subsequent steps will involve conversion between this and the types introduced in #512.

### CODE REVIEW

File: semantic-scope-graph/src/Stack/Path.hs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+
+import Data.Functor.Tagged
+import Data.Monoid
+import Data.Text (Text)
+import Stack.Graph (Node (..), Symbol)
+import Data.Semigroup (sconcat)
+import Data.Sequence (Seq (..))
+
+data Path = Path
```
Comment (joshvera) [inline semantic-scope-graph/src/Stack/Path.hs]: Could we add some 📝 to this? Is the intent to generate `Path` datatypes from the stack graph, then check their validity?

File: semantic-scope-graph/src/Stack/Path.hs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+         Empty
+           -> check startingNode endingNode
+         Edge { sourceNode, sinkNode } :<| Empty
+           -> check startingNode sourceNode <> check endingNode sinkNode
+         Edge { sourceNode  } :<| (_ :|> Edge { sinkNode })
+           -> check startingNode sourceNode <> check endingNode sinkNode
+
+-- | The starting node of a path must be the root node, an exported
+-- scope node, or a reference node. The ending node of a path must be
+-- the root node, a jump to scope node, a resolved jump to scope node,
```
Comment (dcreager) [inline semantic-scope-graph/src/Stack/Path.hs]: > resolved jump to scope node

Gah!  This is a typo in the design doc, "resolved jump to scope node" isn't a type of node anymore.  That clause should be removed.

File: semantic-scope-graph/src/Stack/Path.hs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+checkNodeInvariants Path { startingNode, endingNode }
+  = getFirst (checkStart <> checkEnd)
+    where
+      checkStart = case extract startingNode of
+        Root            -> mempty
+        ExportedScope{} -> mempty
+        Reference{}     -> mempty
+        _other          -> pure (BadStartingNode startingNode)
+
+      -- PT TODO: add the case checking for "resolved" jump to scopes
```
Comment (dcreager) [inline semantic-scope-graph/src/Stack/Path.hs]: This TODO is moot per above

File: semantic-scope-graph/src/Stack/Path.hs (hunk 1 above)
Comment (patrickt) [reply semantic-scope-graph/src/Stack/Path.hs]: I think we want to do a lot more than check their validity, but I wanted to run this design by y’all before we spec out more functionality on the Haskell side.

**Pull request merged**